### PR TITLE
Feature/basis

### PR DESCRIPTION
**Pull request by Jammy2211**
A Basis is a list of light profiles used to model a Galaxy light profile, for example 10 Gaussian's:

```
gaussian_m = af.UniformPrior(lower_limit=0.0, upper_limit=1.0)
gaussian_c = af.UniformPrior(lower_limit=0.0, upper_limit=10.0)

gaussians = af.Collection([af.Model(ag.lp_linear.EllGaussian) for i in range(10)])

for i, gaussian in enumerate(gaussians):

    gaussian.centre = gaussians[0].centre
    gaussian.elliptical_comps = gaussians[0].elliptical_comps
    gaussian.sigma = (gaussian_m * i) + gaussian_c

bulge = af.Model(ag.lp_basis.Basis, light_profile_list=gaussians, regularization=ag.reg.Constant)

galaxy = af.Model(ag.Galaxy, redshift=0.5, bulge=bulge)

model = af.Collection(galaxies=af.Collection(galaxy=galaxy))
```

This PR implements the `Basis` class in the source code. 

A `Basis` can be regularized based on the light profiles it contains. This required the `Inversion` module to be extended to support lists of light profiles being linear solved for (with regularization), which makes up most the work in this PR.

**There is a lot of docs missing, which I will add in a future PR. There is another bigish refactor coming after this PR in the same parts of the code -- but this branch has had a lot of work done to it so I figure its best to get a review done now. **

### CODE REVIEW

File: autoarray/util/misc_util.py [hunk 1]
```diff
@@ -0,0 +1,79 @@
...
+
+    Returns
+    -------
+    The list of objects in the galaxy that inherit from input `cls`.
+    """
+    if cls_filtered is not None:
+        return [
+            value
+            for value in values
+            if isinstance(value, cls) and not isinstance(value, cls_filtered)
```
Comment (rhayes777) [inline autoarray/util/misc_util.py:58]: Might be slightly simpler if you do the cls filter then optionally apply the cls_filtered filter

File: test_autoarray/inversion/inversion/imaging/test_imaging.py [hunk 2]
```diff
@@ -0,0 +1,146 @@
...
+    assert inversion.operated_mapping_matrix[0, 0] == pytest.approx(1.0, 1e-4)
+
+    convolver = aa.m.MockConvolver(operated_mapping_matrix=np.ones((2, 2)))
+
+    inversion = aa.m.MockInversionImaging(
+        convolver=convolver,
+        linear_obj_list=[rectangular_mapper_7x7_3x3, rectangular_mapper_7x7_3x3],
+    )
+
+    operated_mapping_matrix_0 = np.array([[1.0, 1.0], [1.0, 1.0]])
```
Comment (rhayes777) [inline test_autoarray/inversion/inversion/imaging/test_imaging.py:30]: Use fixtures to break these tests up?

File: test_autoarray/inversion/inversion/imaging/test_inversion_imaging_util.py [hunk 3]
```diff
@@ -0,0 +1,340 @@
...
+            pix_pixels=pixelization.pixels,
+        )
+
+        # print(data_vector_via_w_tilde)
+        # print(data_vector)
+
+        assert data_vector_via_w_tilde == pytest.approx(data_vector, 1.0e-4)
+
+
+def test__curvature_matrix_via_w_tilde_two_methods_agree():
```
Comment (rhayes777) [inline test_autoarray/inversion/inversion/imaging/test_inversion_imaging_util.py:240]: These tests could probably be combined using parametrize

File: test_autoarray/inversion/pixelization/test_pixelization_util.py [hunk 4]
```diff
@@ -11,106 +11,97 @@ def test__rectangular_neighbors_from():
...
-    assert (pixel_neighbors[13] == [9, 12, 14, -1]).all()
-    assert (pixel_neighbors[14] == [10, 13, 15, -1]).all()
-    assert (pixel_neighbors[15] == [11, 14, -1, -1]).all()
+    (neighbors, neighbors_sizes) = aa.util.pixelization.rectangular_neighbors_from(
+        shape_native=(4, 4)
+    )
+
+    assert (neighbors[0] == [1, 4, -1, -1]).all()
+    assert (neighbors[1] == [0, 2, 5, -1]).all()
+    assert (neighbors[2] == [1, 3, 6, -1]).all()
```
Comment (rhayes777) [inline test_autoarray/inversion/pixelization/test_pixelization_util.py:88]: Use parametrize to apply each of these assertions

**Review by rhayes777 — APPROVED** [review]
I hope the next PR is shorter...

File: autoarray/inversion/inversion/abstract.py [hunk 5]
```diff
@@ -2,50 +2,50 @@
...
         data
-        leq
-        regularization_list
+        inversion
         settings
         preloads
         profiling_dict
         """
 
+        from autoarray.preloads import Preloads
```
[the remark this answers is not in the document]
Comment (Jammy2211) [reply autoarray/inversion/inversion/abstract.py:45]: Yep.

File: autoarray/inversion/inversion/abstract.py [hunk 6]
```diff
@@ -2,50 +2,50 @@
...
+
+        preloads = preloads or Preloads()
+
         try:
             import numba
```
[the remark this answers is not in the document]
Comment (Jammy2211) [reply autoarray/inversion/inversion/abstract.py:50]: This is already what happens, the error here just completely stops one using `Inversions` without numba as itd be way too slow.

**Pull request merged**